test/test_sbol2_sbol3_direct.py

```diff
@@ -14,140 +14,84 @@
-        doc3 = convert2to3(doc2, use_native_converter=True)
-        self.assertEqual(len(doc3.validate()), 0)
-        with tempfile.TemporaryDirectory() as tmpdir:
-            tmp3 = Path(tmpdir) / 'doc3.nt'
-            doc3.write(tmp3)
-            self.assertFalse(file_diff(str(tmp3), str(TEST_FILES / 'sbol3_implementation.nt')))
-            doc2_loop = convert3to2(doc3, True)
-            # report = doc2.validate()
-            # self.assertEqual(len(report), 0, f'Validation failed: {report}')
-            tmp2 = Path(tmpdir) / 'doc2_loop.xml'
-            doc2_loop.write(tmp2)
-            self.assertFalse(file_diff(str(tmp2), str(TEST_FILES / 'sbol_3to2_implementation.xml')))
-    '''
+        """Test ability to convert an Implementation from SBOL2 to SBOL3"""
+        self.handle_2to3_conversion('sbol_3to2_implementation.xml', 'sbol3_implementation.nt')
 
-    # sbol_3to2_collection.xml is not SBOL compliant. Leaving conversions involving it for after the compliant converter is done
-    '''
-    # sbol_3to2_collection.xml is not SBOL compliant. Leaving conversions involving it for after the compliant converter is done
-    def test_3to2_collection_conversion(self):
-        """Test ability to convert a collection from SBOL3 to SBOL2"""
+    def test_2to3_collection_conversion(self):
+        """Test ability to convert a Collection from SBOL2 to SBOL3"""
+        self.handle_2to3_conversion('sbol_3to2_collection.xml', 'sbol3_collection.nt')
+
+
+class TestDirectSBOL3SBOL2Conversion(unittest.TestCase):
+
+    def handle_3to2_conversion(self, test_filename, rubric_filename):
+        """Provides a re-usable test handler for converting SBOL3 to SBOL2 with different test files
+        test_filename: Name of an SBOL3 test file
+        rubric_filename: Name of an SBOL2 file with expected conversion
+    """
         # Load an SBOL3 document and check its contents
         doc3 = sbol3.Document()
-        doc3.read(TEST_FILES / 'sbol3_collection.nt')
+        doc3.read(TEST_FILES / test_filename)
         # Convert to SBOL2 and check contents
         doc2 = convert3to2(doc3, True)
-        #report = doc2.validate()
-        #self.assertEqual(len(report), 0, f'Validation failed: {report}')
         with tempfile.TemporaryDirectory() as tmpdir:
             tmp2 = Path(tmpdir) / 'doc2.xml'
             doc2.write(tmp2)
-            self.assertFalse(file_diff(str(tmp2), str(TEST_FILES / 'sbol_3to2_collection.xml')))
+            self.assertFalse(file_diff(str(tmp2), str(TEST_FILES / rubric_filename)))
             doc3_loop = convert2to3(doc2, use_native_converter=True)
             self.assertEqual(len(doc3_loop.validate()), 0)
             tmp3 = Path(tmpdir) / 'doc3_loop.nt'
             doc3_loop.write(tmp3)
-            self.assertFalse(file_diff(str(tmp3), str(TEST_FILES / 'sbol3_collection.nt')))
-    '''
+            self.assertFalse(file_diff(str(tmp3), str(TEST_FILES / test_filename)))
+            report = doc2.validate()
```

Comment:
Why has the doc2 conversion moved down from above?
I think that it should be up when we generate the doc2 object, not after we write it.

Reply:
See above.